### ISSUE TITLE
Is it worth creating a centralised CommonResourceHelper Module?

### ISSUE THREAD

**Issue by ld0614** [PowerShell/DscResources#334]
Looking at a number of DSC Modules they have a module either called CommonResourceHelper or have a helper module providing a number of similar functions.  I've also noted that they tend to differ significantly in the way they work and what functions are available, possibly due to divergence from a common fork or simply reimplementation of the same tasks.  

My view is that a version-controlled CommonResourceHelper module would strengthen the DSC Resource Kit by allowing improvements from one module on common areas to be easily shared between modules.  It would also allow common tasks such as localization and event handling to be templatized and possibly even enforced for HQRM modules as they then provide a common way of doing things.  For a new module, new resource being added to a module or simply for other community projects a centralized and tested module would lower the barrier to entry and assist with common friction points.

My initial thoughts for versioning would be to do it in a similar way to the current Unit test templates; that is to provide a version numbered module which each maintainer can then copy into the associated DSC Resource module as they see fit.  This would allow for flexibility while promoting upgradability.  

My initial thoughts on functions to include:
Get-LocalizedData
Start-Process
Wait-ForProcess
New-ErrorMessage
New-WarningMessage
New-VerboseMessage
New-Exception(s)

What are everyone else's thoughts?

**Comment by johlju** [comment]
I all for this idea! I would simplify getting new and existing modules to use the same logic and easier to update those that already do. And also, that there are tests for those functions that can be copied as well.

Instead on `New-Exception` maybe we should use the common among some resource modules, see for examlpe this [CommonResourceHelper.psm1](https://github.com/PowerShell/xSQLServer/blob/dev/DSCResources/CommonResourceHelper.psm1), with the accompanying tests [CommonResourceHelper.Tests](https://github.com/PowerShell/xSQLServer/blob/dev/Tests/Unit/CommonResourceHelper.Tests.ps1) . See also CommonTestHelpers further down. 

`New-InvalidArgumentException`
`New-InvalidOperationException`
`New-ObjectNotFoundException`
`New-InvalidResultException`

`Start-Process` and `Wait-ForProcess` could take many shapes depending on what the resource want or need to do, but if we should have a common one, then it would need to be able to catch errors. The idea of the complexity can be seen in @adbertram's script [Invoke-Process](https://github.com/adbertram/Random-PowerShell-Work/blob/master/Processes/Invoke-Process.ps1). *Not sure if we are allowed to use his script, but I donated him "a few bucks" anyway for writing it and giving a lot to the community. :smile:*

As for the `New-VerboseMessage`, when localizing we should use the below. This is so that the common test (ScriptAnalyzer) do not fail on missing Write-Verbose in the *-TargetResource functions. Although, if there is a need for outputting a prefix or suffix to each verbose message then I can see the need for such a helper function (we have one in xSQLServer which we are moving away from, in favor of `Write-Verbose`).
```
Write-Verbose ($localizedMessage -f $firstVariable, $secondVariable)
``` 

Also, we should consider the same approach for the [CommonTestHelpers](https://github.com/PowerShell/xStorage/blob/dev/Tests/TestHelpers/CommonTestHelper.psm1) which is used to get localized messages in tests.

**Comment by johlju** [comment]
To further add to the complexity with a common Start-Process helper function, I have these in xSQLServer; [Copy-ItemWithRoboCopy](https://github.com/PowerShell/xSQLServer/blob/dev/DSCResources/MSFT_xSQLServerSetup/MSFT_xSQLServerSetup.psm1#L1903) and [Start-SqlSetupProcess](https://github.com/PowerShell/xSQLServer/blob/dev/DSCResources/MSFT_xSQLServerSetup/MSFT_xSQLServerSetup.psm1#L2097). The later could be improve with something like @adbetram's script above that catches error written to the output stream. Today, errors from the SQL setup process are not caught, I have a distant memory of the output being written to the console only and never was redirected to the `RedirectStandardOutput`.

My point is, having a common `Start-Process` helper function might not be easy to maintain.

**Comment by ld0614** [comment]
I'm happy with using the different Exception types currently already in use across the modules, I wasn't aware that xSQLServer was moving away from the `New-VerboseMessage`, are you guys moving away from it for errors and warnings as well?

The idea behind `New-Process` was to allow the system account to launch process correctly, the code is already available in the `Windows-Process` resource in PSDSCResources.  I have already implemented this in a private branch of xSCDPM and I am just waiting on final signoff from my manager before opening up the code.  Basically the `Start-SqlSetupProcess` function would continue to exist as that provides error handling specific to that module, however, the actual starting of the process and potentially the handling of generic errors such as access denied etc are handled by this function.

**Comment by johlju** [comment]
@ld0614 Totally missed to answer you on this. 😞 

SqlServerDsc is moving away from `New-ErrorMessage`, `New-WarningMessage`, and `New-VerboseMessage` mostly because there are a PSSA rule that warns if there isn't a Write-Verbose in each of the Get-,Set-,Test-functions. And the helper functions doesn't provide any benefits.
I personally wouldn't mind using a standard New-Process for the Start-SqlSetupProcess. 

Looking back at was written previously I like the idea of having a standard module with code that can be copied to a resource module, and the first instinct would that it would be added to [DscResource.Template](https://github.com/PowerShell/DscResources/tree/master/DscResource.Template) . The problem I'm seeing with that is that the code can't be tested in this repo, and for adding and changing such code it must be tested.

**Comment by kwirkykat** [comment]
@johlju Yes on adding common code to the template. But I really don't think there should be any code or testing in this repo as this repo is supposed to be for central DSC Resource Kit documentation/discussions only. I would vote for actually making a separate DscResource.Template repo for all the resource module template code and then the common helpers and testing of those helpers could happen there. @Indhukrishna Can you create a new repo for this?

**Comment by johlju** [comment]
@kwirkykat Awesome idea! I like the idea of DscResource.Template! That will be awesome, then we can get a extremly basic resource running with tests and integration tests that is fully working to show of the  basic standard things that all repos can reuse.

**Comment by PlagueHO** [comment]
@kwirkykat , @johlju that idea sounds really great.

**Comment by johlju** [comment]
I have started a DscResource.Template here https://github.com/johlju/DscResource.Template and have reminded @Indhukrishna to create a repo for it. 🙂 

I work on the new repo this weekend, but it is not finished, need a couple of more weekends to finish it. Once finished I will PR it in to the newly created repo (when it exist).

Also, I will use that repo to try to see if it is possible to join the both test framework models discussed in issue https://github.com/PowerShell/DscResource.Tests/issues/225. It would be great if that was part of the template (especially the auto-documentation feature).

**Comment by johlju** [comment]
Still working on this, but slow progress as I do this when I have time. We have yet any repo to publish it to either.

**Comment by kwirkykat** [comment]
@johlju
https://github.com/PowerShell/DscResource.Template

**Comment by johlju** [comment]
@kwirkykat awesome thank you 😄

**Comment by PlagueHO** [comment]
Nice!!!

**Comment by gaelcolas** [comment]
I think the `DscResource.template` and a `resource helper module` of common functions should be two separate things (eventually).
Resource modules wanting to leverage the Resource Helper Module can then pull that module from the gallery, pin a version of their liking, or always pull the latest (and see if a new version breaks their build).

Not using Modules as artefacts is bad practice, and modules created with a template at a given time, is unlikely to be ever updated (and it's very hard to track).
Using module dependency / nested modules, that pulls artefacts at build time (similar to nuget restore) is the right way to do it.

It should facilitate iteration (smaller batch size), testing, and composition of loosely coupled modules.

**Comment by johlju** [comment]
@gaelcolas You mean that resource module should be dependent of the "Resource Helper Module" in the `RequiredModules` element of the module manifest? It should be automatically downloaded from the Gallery when the user does, for example, `Install-Module`?
If we go in that direction then I think we can leverage the same repo for both those separate things. 🤔

**Comment by gaelcolas** [comment]
> "You mean that resource module should be dependent of the "Resource Helper Module" in the `RequiredModules` element of the module manifest?"

Yes, that's one way of doing it

> "It should be automatically downloaded from the Gallery when the user does, for example, `Install-Module`?"

Yes, but you could technically 'embed' it into your module (although I don't think that's a good idea).
The important part is to reference an external tested and trusted artefact, managed in another repository and pipeline.

> "If we go in that direction then I think we can leverage the same repo for both those separate things"

Technically possible, but risky.
The `DscResource.Template` should just be the scaffolding of a resource module including, say, a plaster template, build files, and so on. 
That template should have tests and look like a fully fledged resource module, build on AppVeyor and, all in all, be a reference. As such, it would reference that "Resource Helper module" from the Gallery (but not contain its source code).

Sure we could bundle them in one repo, but that **could get akward**:
- You don't need Plaster template files and tests in your "Resource Helper module" you use with many Resource module
- You need to be able to change one without changing the other (smaller change domain, less revision, different release cycle/cadence)

Hope that makes sense!

**Comment by PlagueHO** [comment]
There are a few things to consider with having a shared "Resource Helper Module" in PS Gallery:
1. We'd need to ensure that we were using new and unique function names in the new shared "Resource Helper Module" - otherwise we'd end up with nasty naming conflicts (like when I tried including the ResourceHelper.psm1 in all the modules I support and it broke them all).
2. If you're using DSC with Pull mode or if the target node doesn't have internet access then you'd need to make sure you're packaging and deploying that dependent module to all your nodes. From memory (I could be wrong), but the target node won't pull down any "RequiredModules" onto the target node for any DSC Resources consumed in a config. And many DSC target nodes don't necessarily have direct access to the PS Gallery.

So, IMHO, although I am fully in support of using immutable artifact principles here, it might be better to shift the dependency to managed at build time rather than run time.

Perhaps we could still package and store the shared code in the PSGallery, but pull it down and include it at build/test time.

**Comment by gaelcolas** [comment]
@PlagueHO yes what I meant by nested is to cater for your point 2. You exactly got my idea.

You reference, say `DscCommonHelpers` version `1.3.2` in your Dependency file, and it just add from the gallery, to your build output module, as a nested module (you just don't add it to your Git repo, you rely on the artefact).

**Comment by PlagueHO** [comment]
Cool! That sounds like a workable solution there!

**Comment by johlju** [comment]
We can't reply on the Gallery for this. Is it possible for it to fetch from a private gallery too? Because all users might not allow fetching production code directly from the public Gallery. 🤔 

I like the idea discussed above, but does anyone have the time to put in the work if we would get a another repo 'DscResource.Common' from @kwirkykat? 
We could automatically deploy the repo using the same account that deploys the examples to the Gallery.

My current though is the DscResource.Template will be a fully functioning module with fully working tests etc. to make sure the common code works. It will also have the scaffolding for tests templates, resource etc. beside that. I have that almost finished already, I think (was working on this way way back). Even if we go with the 'DscResource.Common' we need somewhere to integration test that, this DscResource.Template would work for that?

**Comment by gaelcolas** [comment]
Private Gallery for this would work, definitely. The principle would also work with the `DscResource.tests`, better than git clone in secure environment.

Actually this is what I've done for the last few years ([see my SampleModule](https://gaelcolas.com/2017/10/22/a-sample-powershell-module-pipeline-with-plaster-template/) project, which I'm improving with [ModuleBuilder](https://github.com/PoshCode/ModuleBuilder) right now. 

So it wouldn't be too hard to merge with your work.

That said, it should not be a blocker (hence the 'eventually'), and you should definitely push your work to `DscResource.Template`.
And yes again, it would be the right place for Integration testing!

**Comment by johlju** [comment]
Yes let us move the discussion over to the DscResource.Template (thanks for opening another issue there  @gaelcolas when you have time). 
Once I merged the code into DscResource.Template, and have removed the scaffolding from this repo we should close this issue. 

I will work on getting the scaffolding out of this repo.
